File: problem023/solution.py
from src import helpers


def divisors(num):
    divisor = [1]
    x = 2
    while 2 * x <= num:
        if num % x == 0:
            divisor.append(x)
        x += 1
    return divisor


def is_abundant(num):
    return sum(divisors(num)) > num


# Calculate list of abundant numbers below threshold
abundants = [x for x in range(1, 28124) if is_abundant(x)]

# Calculate all sums of two abundant numbers
sums = set()
for x in abundants:
    for y in abundants:
        sums.add(x + y)

nonabundant_sums = [x for x in range(1, 28124) if x not in sums]

helpers.solution(sum(nonabundant_sums))

File: problem023/solution.php
<?php
$start = microtime(true);
require __dir__ . "/../src/helpers.php";

function divisors($num) : array
{
	for($divs = [1], $x = 2; 2 * $x <= $num; $x++) {
		if ($num % $x == 0) {
			$divs[] = $x;
		}
	}
	return $divs;
}

function isAbundant($num) : bool 
{
	return array_sum(divisors($num)) > $num;
}

// Find the Abundant Numbers Below Threshold
for ($i = 1; $i < 28123; $i++) {
	if(isAbundant($i)) {
		$abundants[] = $i;
	}
}

// Find Abundant Sums
foreach ($abundants as $a) {
	foreach ($abundants as $b) {
		$sums[$a + $b] = true;
	}
}

// Find Total of Non-abundant sums
for($j = 1; $j <= 28123; $j++) {
	if(!isset($sums[$j])) {
		$total[] = $j;
	}
}

solution(array_sum($total));
echo "Duration: ", microtime(true) - $start, PHP_EOL;

